python/lsst/pipe/tasks/characterizeImage.py

```diff
@@ -0,0 +1,493 @@
+        - sourceCat: detected sources (an lsst.afw.table.SourceCatalog)
+        - background: model of background subtracted from exposure (an lsst.afw.math.BackgroundList)
+        - psfCellSet: spatial cells of PSF candidates (an lsst.afw.math.SpatialCellSet)
+        """
+        self._frame = self._initialFrame # reset debug display frame
+
+        if not self.config.doMeasurePsf and not exposure.hasPsf():
+            raise pipeBase.TaskError("exposure has no PSF model and config.doMeasurePsf false")
+
+        if background is None:
+            background = BackgroundList()
+
+        # make a deep copy of the mask
+        originalMask = exposure.getMaskedImage().getMask().clone()
+
+        # subtract an initial estimate of background level
+        estBg = estimateBackground(
+            exposure = exposure,
+            backgroundConfig = self.config.background,
+            subtract = False, # this makes a deep copy, which we don't want
+        )[0]
+        image = exposure.getMaskedImage().getImage()
+        image -= estBg.getImageF()
+        background.append(estBg)
+
+        psfIterations = self.config.psfIterations if self.config.doMeasurePsf else 1
+        for i in range(psfIterations):
+            if i > 1:
+                # restore original mask so that detections and cosmic rays
+                # are only marked by the final iteration
+                exposure.getMaskedImage().getMask()[:] = originalMask
+
+            dmeRes = self.detectMeasureAndEstimatePsf(
+                exposure = exposure,
+                exposureIdInfo = exposureIdInfo,
+                background = background,
+            )
+
+            psf = dmeRes.exposure.getPsf()
+            psfSigma = psf.computeShape().getDeterminantRadius()
+            psfDimensions = psf.computeImage().getDimensions()
+            medBackground = np.median(dmeRes.background.getImage().getArray())
+            self.log.info("iter %s; PSF sigma=%0.2f, dimensions=%s; median background=%0.2f" % \
+                (i + 1, psfSigma, psfDimensions, medBackground))
+
+        self.display("psf", exposure=dmeRes.exposure, sourceCat=dmeRes.sourceCat)
+
+        # perform final repair with final PSF
+        self.repair.run(exposure=dmeRes.exposure)
+        self.display("repair", exposure=dmeRes.exposure, sourceCat=dmeRes.sourceCat)
+
+        # perform final measurement with final PSF, including measuring and applying aperture correction,
+        # if wanted
+        self.detectAndMeasure.measure(
+            exposure = dmeRes.exposure,
+            exposureIdInfo = exposureIdInfo,
+            sourceCat = dmeRes.sourceCat,
+            allowApCorr = True,  # the default; listed for emphasis
+        )
+        self.display("measure", exposure=dmeRes.exposure, sourceCat=dmeRes.sourceCat)
+
+        return dmeRes
+
+    @pipeBase.timeMethod
+    def detectMeasureAndEstimatePsf(self, exposure, exposureIdInfo, background):
```

Comment:
Why is this called "EstimatePsf", when in other places we refer to measuring the PSF?
I think naming this method "iterate" or "iterateOnce" might make it a bit easier.